_posts/2016-09-30-fiber-bomb-debugging-story.md

```diff
@@ -0,0 +1,66 @@
+---
+layout: post
+title: "The Mysterious Fiber Bomb Problem: A Debugging Story"
+author: Kenton Varda
+authorUrl: https://github.com/kentonv
+---
+
+A month or two ago, we started seeing a mysterious problem in production: every now and then, one of our Node.js web server processes supporting Sandstorm Oasis would suddenly jump to 100% CPU usage (of one core) and stay there until it was killed. The problem wasn't an infinite loop, though: the process continued to respond to requests, just slowly. Since the process continued to respond to requests, it continued to pass health checks and was never restarted automatically. But for users assigned to that shard, the service was essentially unusable, as every action would take seconds to complete. The problem left nothing at all suspicious in the logs -- other than a gap in which far fewer requests that normal were being handled. At first, the problem only struck about once a week, seemingly at random.
+
+This kind of bug is a web developer's worst nightmare. How do you debug something which you can only reproduce once a week, at random, with real users on the line? What could even cause a process to slow down but not stop in this way?
+
+### What's eating our CPU?
+
+Obviously, we needed to take a CPU profile while the bug was in progress. Of course, the bug only reproduced in production, therefore we'd have to take our profile in production. This ruled out any profiling technology that would harm performance at other times -- so, no instrumented binaries. We'd need a sampling profiler that could run on an existing process on-demand. And it would have to understand both C++ and V8 Javascript. (This last requirement ruled out my personal favorite profiler, pprof from google-perftools.)
+
+Luckily, it turns out there is a correct modern answer: Linux's "perf" tool. This is a sampling profiler that relies on Linux kernel APIs, thus not requiring loading any code into the target binary at all, at least for C/C++. And for Javascript, it turns out v8 has built-in support for generating a "perf map", which tells the tool how to map JITed code locations back to Javascript source: just pass the `--perf_basic_prof_only_functions` flag on the Node command-line. This flag is safe in production -- it writes some data to disk over time, but we rebuild all our VMs weekly, so the files never get large enough to be a problem.
```

Comment:
This is probably a dumb one, but v8 -> V8 for consistency with the rest of the post?